### DISCUSSION TITLE
Users can opt-in to userclasses but cannot opt-out of them

### DISCUSSION THREAD

**Discussion by Jimmi08** [e107inc/e107#5162] [Q&A Core]
Moving my posts here from to not mess with issue 

https://github.com/e107inc/e107/issues/5159  

**Bug description**
Once a user has assigned themselves any custom userclass on the settings page, they cannot unassign it. In the database, I have noticed in the "e107_user" table that the "user_class" column is completely overwritten with the id of the userclass instead of it being appended to it. This happens on my current production install and a fresh test install.

I can confirm this in 7.4 too. And latest github version.

If you want a quick fix, I did it because I am using it on the live site too. But I have no idea why it helped and what is a problem, ok?  Tested on e107sk.com

![image](https://github.com/e107inc/e107/assets/5429548/b3f01d38-fe79-4b3f-9f82-ef6a707681e8)

On left side is fix, on right side is actual version of handlers/user_class.php
 
```
		$e107 = e107::getInstance();
		$ecache = $e107->getCache();
```

> Unfortunately, nothing has changed after applying your fix and clearing the cache. 

I am sorry, it works for me - PHP 7.4 - locally with 8.1.  I am not able to help more,

**Comment by Jimmi08** [comment]
@HatchlingByHeart  delete your answer too if you can.  Sorry for messing with your issue (it is better to left it untouched and discuss here)

**Reply by HatchlingByHeart** [reply]
Done.

**Comment by Alex-e107nl** [comment]
It is changed on 0ct 27.. [https://github.com/e107inc/e107/commit/552af19c54dabb7f037ec6f27d5652279f600c30](https://github.com/e107inc/e107/commit/552af19c54dabb7f037ec6f27d5652279f600c30) PHP 8.2 Deprecated error removal ?

I can confirm the problem, today's git....

**Reply by Jimmi08** [reply]
Yeah, it was that commit. And because it is part of 2.3.3 release, I tried to fix it and it worked for me.  

Do you have an issue that you can't change access for userclasses (manage and visible) too? Because for me it is working.

**Reply by Alex-e107nl** [reply]
The permissions change automatically, I set it as everyone (public) but it keeps going to No One (inactive) Manager goes to Admin
Visibility goes to admin Does not matter what I choose... (should be members / members I think) I'm getting a bit confused :-)

**Reply by Jimmi08** [reply]
So it is not related to my fix.  @HatchlingByHeart  mentioned this but during the moving discussion here, it was deleted (I wrote delete instead move it here).  It should be created as the normal issue.

**Reply by HatchlingByHeart** [reply]
I did mention that I could not even change the permissions on the userclasses either. Since I did not notice this before applying Jimmi08's fix, I don't know if his change caused it or not, but given Alex's response, it seems unlikely.

**Reply by HatchlingByHeart** [reply]
Also, I am on PHP 8.2.13, which to my knowledge is the latest on the PHP 8.2.x branch.

Would downgrading to 8.1 or upgrading to 8.3 possibly help? @Jimmi08 @Alex-e107nl

**Reply by Jimmi08** [reply]
No idea, but I think that e107 fully supports 8.1. That commit is called PHP 8.2 fixes but it is changing other things too.  

There is time zones issue and it is holiday time...   you need to wait for core developers with this.

**Reply by Alex-e107nl** [reply]
I am also working on 8.1.26 so that's no solution... keep in mind that not everyone is in this timezone... wait and see if one of the dev's have a solution?

**Reply by Jimmi08** [reply]
I am not sure what to test.  I have 2.3.3 installation, some plugins are there, custom theme... Userclasses work normally. PHP 8,2.

![image](https://github.com/e107inc/e107/assets/5429548/40937226-f9a8-4489-9c19-4e2eef1f535c)

**Reply by HatchlingByHeart** [reply]
Given that the issue also happened on a fresh install of 2.3.3 for me, I can rule out any changes to files I have made. My theory is it may be related to PHP config. That's the only thing I can imagine would differ between a fresh install on my system and someone else's.

My server OS is Debian 11, and my PHP config can be found here (comments removed): https://www.dropbox.com/scl/fi/lmp1kip6mcjmhn0ufac3z/php.ini?rlkey=uqw3zy6p8s3pcjuavh20lpu4o&dl=0

Can you guys think of anything else that may differ in our environments?

**Comment by HatchlingByHeart** [comment]
For anyone curious about my process in setting up when I get these issues, here's the video I posted in the issue.

https://github.com/e107inc/e107/assets/6427368/9b94b8c8-6be5-407a-b649-18b768e06ef2

**Comment by HatchlingByHeart** [comment]
Another user appears to have found the problem.

Quote (@Deltik): (#5159)
```
I was able to reproduce the bug as a normal user (not admin) by unchecking all the "Subscribed to" items.

The cause is that the class[] array being empty is not sent to the server for validation, so the server assumes that the client did not want to change the userclasses.

As for the "Manager" and "Visibility" changing from "Everyone (public)" to "Admin" in /e107_admin/userclass2.php?mode=main&action=edit, it's because "Everyone (public)" is saved as 0, which is interpreted as no value, so the default will be "Admin" when the page loads again. It is actually saved as "Everyone (public)", but a subsequent save will actually change it to "Admin".

So there are two bugs.
```

**Comment by HatchlingByHeart** [comment]
@Jimmi08 @Alex-e107nl
I was unable to fix the code, but I have a workaround that works:

Adding a dummy userclass that is otherwise not used for anything (ID# 1 in my case), and making the following changes:

**/e107_core/shortcodes/batch/userclasses_shortcodes.php**
![Screenshot 2023-12-23 074609](https://github.com/e107inc/e107/assets/6427368/4b3ec5dc-04e7-4139-9b1f-bbbf01e062ad)

**/e107_handlers/userclass_class.php**
![Screenshot 2023-12-23 074649](https://github.com/e107inc/e107/assets/6427368/fafe7950-a556-4fd9-b69c-94a3aff5739a)

**Comment by Vodhin** [comment]
I came across this issue a while ago. I added a bit of code to usersettings.php around line 540 to ensure that userclass was never empty by adding a useless value;
```

// Update Userclass - only if its the user changing their own data (admins can do it another way)
$allData['data']['user_class'] = ($allData['data']['user_class'] ? $allData['data']['user_class'] : -256);
if (isset($allData['data']['user_class']))
{ ...
```

this seems to work for me.
